targets/XPlatCppSdk/source/code/source/playfab/PlayFabAndroidHttpPlugin.cpp

```diff
@@ -0,0 +1,743 @@
+    {
+    public:
+        RequestImpl() :
+            jniEnv(nullptr),
+            httpRequestObject(nullptr)
+        {
+
+        }
+
+        virtual ~RequestImpl()
+        {
+            if (this->httpRequestObject)
+            {
+                assert(this->jniEnv);
+                this->jniEnv->DeleteGlobalRef(this->httpRequestObject);
+                this->httpRequestObject = nullptr;
+            }
+            this->jniEnv = nullptr;
+        }
+
+        void SetJniEnv(JNIEnv* jniEnv)
+        {
+            this->jniEnv = jniEnv;
+        }
+
+        JNIEnv* JniEvn()
+        {
+            assert(this->jniEnv);
+            return this->jniEnv;
+        }
+
+        jobject GetHttpRequestObject()
+        {
+            if (this->httpRequestObject == nullptr)
+            {
+                assert(this->jniEnv);
+                assert(GetHelper().GetHttpRequestClass());
+
+                jmethodID initMethod = this->jniEnv->GetMethodID(GetHelper().GetHttpRequestClass(), "<init>", "()V");
+                assert(initMethod);
+                if (initMethod == nullptr)
+                {
+                    return nullptr;
+                }
+                jobject localObject = this->jniEnv->NewObject(GetHelper().GetHttpRequestClass(), initMethod);
+                assert(localObject);
+                if (localObject == nullptr)
+                {
+                    return nullptr;
+                }
+
+                this->httpRequestObject = this->jniEnv->NewGlobalRef(localObject);
+                assert(this->httpRequestObject);
+
+                this->jniEnv->DeleteLocalRef(localObject);
+            }
+            return this->httpRequestObject;
+        }
+
+    private:
+        JNIEnv* jniEnv;
+        jobject httpRequestObject;
+
+    };
+
+}
+
+namespace PlayFab
+{
+    PlayFabAndroidHttpPlugin::RequestTask::RequestTask() :
+        state(RequestTask::None),
+        impl(nullptr),
+        requestContainer(nullptr)
+    {
+
+    }
+
+    PlayFabAndroidHttpPlugin::RequestTask::~RequestTask()
+    {
+        if(impl)
+        {
+            delete impl;
+            impl = nullptr;
+        }
+    }
+
+    bool PlayFabAndroidHttpPlugin::RequestTask::Initialize(std::unique_ptr<CallRequestContainerBase>& requestContainer)
+    {
+        this->requestContainer = std::move(requestContainer);
+        this->impl = new PlayFabAndroidHttpPlugin::RequestImpl();
+        return true;
+    }
+}
+
+namespace PlayFab
+{
+    PlayFabAndroidHttpPlugin::PlayFabAndroidHttpPlugin() :
+        workerThread(nullptr),
+        threadRunning(false),
+        requestingTask(nullptr)
+    {
+
+    }
+
+    PlayFabAndroidHttpPlugin::~PlayFabAndroidHttpPlugin()
+    {
+        threadRunning = false;
+
+        httpRequestMutex.lock();
+
+        if(workerThread)
+        {
+            httpRequestMutex.unlock();
+            workerThread->join();
+            httpRequestMutex.lock();
+        }
+        httpRequestMutex.unlock();
```

Comment:
can we just lock once ?

Reply:
No, to call join method, the lock shouldn't be locked or this will be make dead lock. because the thread is trying to lock during working.

---
In reply to: [274690883](https://github.com/PlayFab/SDKGenerator/pull/470#discussion_r274690883) [](ancestors = 274690883)